### PR TITLE
[leaf_function][print] Add register_multi_grad_hook for leaf function and debug print backwards and multi-tensors

### PR DESCRIPTION
**Pull request by fxdawnn**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #179609

## Summary

Adds `register_multi_grad_hook` API to `@leaf_function` and a new
`torch.utils.debug_log.debug_grad_log` utility built on top of it.

### `register_multi_grad_hook` on leaf functions

`@leaf_function` now supports an optional `@fn.register_multi_grad_hook`
decorator that registers a backward hook. The hook fires exactly once per
backward pass when all `requires_grad` tensor inputs have their gradients
computed. The hook receives the gradients in place of the original tensors
(non-tensor args and non-requires_grad tensors are excluded). The hook itself
runs as a leaf function, so it is opaque to the compiler.

```python
@leaf_function
def my_fn(x, y):
    return (x * 2 + y * 3,)

@my_fn.register_fake
def my_fn_fake(x, y):
    return (torch.empty_like(x),)

@my_fn.register_multi_grad_hook
def my_fn_hook(x_grad, y_grad):
    print(f"x_grad norm: {x_grad.norm()}, y_grad norm: {y_grad.norm()}")
```

### `torch.utils.debug_log.debug_grad_log`

A compile-safe utility for logging gradient norms during backward. No-op in
forward. During backward, logs gradient norms for all monitored tensors:

```python
from torch.utils.debug_log import debug_grad_log

x = torch.randn(4, requires_grad=True)
y = torch.randn(4, requires_grad=True)
z = x * 2 + y * 3

debug_grad_log(x, y)  # no assignment needed

z.sum().backward()
# Logs: [rank 0][bwd] t0_grad_norm=... t1_grad_norm=...
```

## Changes

- `torch/_dynamo/decorators.py`: Add `register_multi_grad_hook` setter on
  leaf functions; thread `hook_fn`/`hook_fake_fn` through
  `_invoke_leaf_function_python`
- `torch/_higher_order_ops/invoke_leaf_function.py`: In
  `InvokeLeafFunctionAutogradOp.forward`, register
  `torch.autograd.graph.register_multi_grad_hook` on requires_grad inputs
  when a hook is present
- `torch/utils/debug_log.py`: New module with `debug_grad_log` — a leaf
  function with a multi-grad hook that logs gradient norms
- `test/functorch/test_leaf_function.py`: 9 new tests for the hook API
  covering single/multi tensor, non-tensor args, no-requires-grad,
  side-effect-only functions, gradient correctness, downstream computation,
  and retain_graph
- `test/higher_order_ops/test_debug_log.py`: 5 tests for `debug_grad_log`

## Test plan

```
python test/functorch/test_leaf_function.py TestLeafFunctionRegisterHook -v
python test/higher_order_ops/test_debug_log.py -v
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98 @mlazos

### CODE REVIEW

File: torch/_higher_order_ops/invoke_leaf_function.py [hunk 1]
```diff
@@ -705,6 +705,49 @@ def fake_backward(*grads):
...
+            if grad_indices:
+                grad_tensors = [flat_args[i] for i in grad_indices]
+
+                def _multi_grad_callback(
+                    grads: Sequence[torch.Tensor | None],
+                ) -> None:
+                    new_flat_args = list(flat_args)
+                    for idx, g in zip(grad_indices, grads):
+                        if g is not None:
+                            new_flat_args[idx] = g
```
Comment (ydwu4) [inline torch/_higher_order_ops/invoke_leaf_function.py]: why do we need to re-set the "new_flat_args[idx] = g"? does it work if we just use flat_args?

File: test/functorch/test_leaf_function.py [hunk 2]
```diff
@@ -2475,5 +2476,297 @@ def forward(self, x):
...
+
+        x = torch.randn(4, requires_grad=True)
+        aot_f = aot_function(f, fw_compiler=nop, bw_compiler=nop)
+        out = aot_f(x)
+        out.sum().backward()
+
+        self.assertEqual(len(hook_grads), 1)
+        self.assertEqual(hook_grads[0], torch.full((4,), 3.0))
+
+    @unittest.expectedFailure
```
Comment (ydwu4) [inline test/functorch/test_leaf_function.py]: what's going on here?

File: test/functorch/test_leaf_function.py [hunk 3]
```diff
@@ -2475,5 +2476,297 @@ def forward(self, x):
...
+        def my_fn_hook(x_grad):
+            hook_count[0] += 1
+
+        x = torch.randn(3, requires_grad=True)
+        out = my_fn(x)[0]
+        # First backward works
+        out.sum().backward()
+        self.assertEqual(hook_count[0], 1)
+
+    @unittest.expectedFailure
```
Comment (ydwu4) [inline test/functorch/test_leaf_function.py]: same here

File: test/functorch/test_leaf_function.py [hunk 4]
```diff
@@ -2475,5 +2476,297 @@ def forward(self, x):
...
+        x = torch.randn(3, requires_grad=True)
+        out = my_fn(x, "hello", 5.0)[0]
+        out.sum().backward()
+
+        self.assertEqual(captured["tag"], "hello")
+        self.assertEqual(captured["scale"], 5.0)
+        self.assertEqual(captured["grad"], torch.full((3,), 5.0))
+
+    def test_hook_multiple_tensor_inputs(self):
+        """Hook fires exactly once with all gradients substituted."""
```
Comment (ydwu4) [inline test/functorch/test_leaf_function.py]: Can remove the comments in the tests? the test is pretty clear.

File: test/higher_order_ops/test_debug_log.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        )
+
+        bwd = [r for r in capture.records if "[bwd]" in r]
+        self.assertEqual(len(bwd), 1)
+        # d(x*2+y*3)/dx = 2, d(x*2+y*3)/dy = 3
+        self.assertIn("t0_grad_norm=2.0000", bwd[0])
+        self.assertIn("t1_grad_norm=3.0000", bwd[0])
+
+    def test_no_requires_grad_no_log(self):
+        """No backward log when no tensor requires grad."""
```
Comment (ydwu4) [inline test/higher_order_ops/test_debug_log.py]: remove the test comments?

File: torch/_higher_order_ops/invoke_leaf_function.py (hunk 1 above)
Comment (fxdawnn) [reply torch/_higher_order_ops/invoke_leaf_function.py]: It's tricky if we were to pass in tag in formation (e.g., rank info). But i've dropped those so we don't need the mapping of arg anymore.

File: test/functorch/test_leaf_function.py (hunk 2 above)
Comment (fxdawnn) [reply test/functorch/test_leaf_function.py]: removed. thanks!

File: test/higher_order_ops/test_debug_log.py (hunk 5 above)
Comment (fxdawnn) [reply test/higher_order_ops/test_debug_log.py]: sounds good. shortened!

File: torch/_higher_order_ops/invoke_leaf_function.py [hunk 6]
```diff
@@ -705,6 +705,41 @@ def fake_backward(*grads):
...
+                arg
+                for arg in flat_args
+                if isinstance(arg, torch.Tensor) and arg.requires_grad
+            ]
+            if grad_tensors:
+
+                def _multi_grad_callback(
+                    grads: Sequence[torch.Tensor | None],
+                ) -> None:
+                    flat_grads = [g for g in grads if g is not None]
```
Comment (ydwu4) [inline torch/_higher_order_ops/invoke_leaf_function.py]: Do we need this filtering? When g is None? My mental model is that every tensor in grad_tensors should get a tensor gradient.

**Review by ydwu4 — COMMENTED** [review]
Seems a test failure?

File: torch/_higher_order_ops/invoke_leaf_function.py (hunk 6 above)
Comment (fxdawnn) [reply torch/_higher_order_ops/invoke_leaf_function.py]: i guess we don't need it. Updated! Thank you!

**Review by ydwu4 — APPROVED** [review]
